fluent-bundle/src/bundle.rs

```diff
@@ -134,6 +249,74 @@ impl<'bundle> FluentBundle<'bundle> {
         }
     }
 
+    /// Formats the message value identified by `path` using `args` to
+    /// provide variables. `path` is either a message id ("hello"), or
+    /// message id plus attribute ("hello.tooltip").
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fluent_bundle::bundle::FluentBundle;
+    /// use fluent_bundle::resource::FluentResource;
+    /// use fluent_bundle::types::FluentValue;
+    /// use std::collections::HashMap;
+    ///
+    /// let resource = FluentResource::try_new("intro = Welcome, { $name }.".to_string()).unwrap();
+    /// let mut bundle = FluentBundle::new(&["en-US"]);
+    /// bundle.add_resource(&resource);
+    ///
+    /// let mut args = HashMap::new();
+    /// args.insert("name", FluentValue::from("Rustacean"));
+    ///
+    /// let value = bundle.format("intro", Some(&args));
+    /// assert_eq!(value, Some(("Welcome, Rustacean.".to_string(), vec![])));
+    ///
+    /// ```
+    ///
+    /// An example with attributes and no args:
+    ///
+    /// ```
+    /// use fluent_bundle::bundle::FluentBundle;
+    /// use fluent_bundle::resource::FluentResource;
+    ///
+    /// let resource = FluentResource::try_new("
+    /// hello =
+    ///     .title = Hi!
+    ///     .tooltip = This says 'Hi!'
+    /// ".to_string()).unwrap();
+    /// let mut bundle = FluentBundle::new(&["en-US"]);
+    /// bundle.add_resource(&resource);
+    ///
+    /// let value = bundle.format("hello.title", None);
+    /// assert_eq!(value, Some(("Hi!".to_string(), vec![])));
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// If no message is found at `path`, then `format` returns `None`.
+    /// 
+    /// In all other cases, `format` returns a string even if it
+    /// encountered errors. `format` uses two fallback techniques to
+    /// create the fallback string. If there are bad references in the
+    /// message, then they will be substituted with `'___'`. If there
+    /// are more extensive errors, then `format` will fall back to using
+    /// `path` itself as the formatted string. Sometimes, but not always,
+    /// these partial failures will emit extra error information in the
+    /// second term of the return tuple.
```

Comment:
If we don't collect any of the errors encountered in the production of the string, it's an implementation shortcoming and I'd prefer not to document it as it indicates the design.

In 0.6 we'll be completing error reporting.

For that reason I'd prefer this sentence to say "Errors produced during message formatting are accumulated in the second tuple field of the Result::Err" or sth like this.